File: springboot/src/test/java/io/proofmark/showad/AccessPolicyEvaluatorTest.java
package io.proofmark.showad;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;
import org.springframework.mock.web.MockHttpServletRequest;

import io.proofmark.showad.access.AccessPolicy;
import io.proofmark.showad.access.AccessPolicyEvaluator;

class AccessPolicyEvaluatorTest {

    private final AccessPolicyEvaluator evaluator = new AccessPolicyEvaluator();

    @Test
    void disabledPolicyContinues() {
        ShowAdProperties.AccessPolicy config = new ShowAdProperties.AccessPolicy();
        config.setEnabled(false);

        AccessPolicy decision = evaluator.evaluate(new MockHttpServletRequest(), config);
        assertThat(decision.action()).isEqualTo(AccessPolicy.Action.CONTINUE);
    }

    @Test
    void cidrAllowlistAllows() {
        ShowAdProperties.AccessPolicy config = new ShowAdProperties.AccessPolicy();
        config.setEnabled(true);
        config.setAllowCidrs(List.of("10.0.0.0/8"));
        config.setTrustedIpHeaders(List.of("X-Forwarded-For"));

        MockHttpServletRequest request = new MockHttpServletRequest();
        request.addHeader("X-Forwarded-For", "10.5.5.5, 1.2.3.4");

        AccessPolicy decision = evaluator.evaluate(request, config);
        assertThat(decision.action()).isEqualTo(AccessPolicy.Action.ALLOW);
        assertThat(decision.reason()).isEqualTo("cidr_allowlist");
    }

    @Test
    void uaAloneDoesNotBypass() {
        ShowAdProperties.AccessPolicy config = new ShowAdProperties.AccessPolicy();
        config.setEnabled(true);
        config.getCrawler().setEnabled(true);
        config.getCrawler().setFamilyCidrs(Map.of("google", List.of("66.249.64.0/19")));

        MockHttpServletRequest request = new MockHttpServletRequest();
        request.addHeader("User-Agent", "Googlebot/2.1");
        request.setRemoteAddr("203.0.113.5");

        AccessPolicy decision = evaluator.evaluate(request, config);
        assertThat(decision.action()).isEqualTo(AccessPolicy.Action.CONTINUE);
    }

    @Test
    void verifiedCrawlerCidrAllows() {
        ShowAdProperties.AccessPolicy config = new ShowAdProperties.AccessPolicy();
        config.setEnabled(true);
        config.getCrawler().setEnabled(true);
        config.getCrawler().setFamilyCidrs(Map.of("google", List.of("66.249.64.0/19")));

        MockHttpServletRequest request = new MockHttpServletRequest();
        request.addHeader("User-Agent", "Mozilla/5.0 (compatible; Googlebot/2.1; +http://google.com)");
        request.setRemoteAddr("66.249.66.1");

        AccessPolicy decision = evaluator.evaluate(request, config);
        assertThat(decision.action()).isEqualTo(AccessPolicy.Action.ALLOW);
        assertThat(decision.reason()).isEqualTo("crawler:google");
    }

    @Test
    void cloudflareVerifiedBotAllowsWithUaMatch() {
        ShowAdProperties.AccessPolicy config = new ShowAdProperties.AccessPolicy();
        config.setEnabled(true);
        config.getCrawler().setEnabled(true);
        config.getCrawler().setAllowCloudflareVerifiedBot(true);

        MockHttpServletRequest request = new MockHttpServletRequest();
        request.addHeader("User-Agent", "Bingbot");
        request.addHeader("CF-Verified-Bot", "true");
        request.setRemoteAddr("172.69.1.1");

        AccessPolicy decision = evaluator.evaluate(request, config);
        assertThat(decision.action()).isEqualTo(AccessPolicy.Action.ALLOW);
        assertThat(decision.reason()).isEqualTo("crawler:bing");
    }

    @Test
    void resolveClientIpPrefersTrustedHeader() {
        ShowAdProperties.AccessPolicy config = new ShowAdProperties.AccessPolicy();
        config.setTrustedIpHeaders(List.of("X-Forwarded-For"));
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.addHeader("X-Forwarded-For", "1.1.1.1, 2.2.2.2");
        request.setRemoteAddr("9.9.9.9");

        String ip = evaluator.resolveClientIp(request, config.getTrustedIpHeaders());
        assertThat(ip).isEqualTo("1.1.1.1");
    }

    @Test
    void defaultTrustedIpHeadersDoNotTrustForwardedHeaders() {
        ShowAdProperties.AccessPolicy config = new ShowAdProperties.AccessPolicy();
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.addHeader("X-Forwarded-For", "1.1.1.1");
        request.setRemoteAddr("9.9.9.9");

        String ip = evaluator.resolveClientIp(request, config.getTrustedIpHeaders());
        assertThat(ip).isEqualTo("9.9.9.9");
    }

    @Test
    void resolveClientIpFallsBackToRemoteAddr() {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRemoteAddr("9.9.9.9");

        String ip = evaluator.resolveClientIp(request, List.of("X-Forwarded-For"));
        assertThat(ip).isEqualTo("9.9.9.9");
    }
}

File: php/tests/AccessPolicyEvaluatorTest.php
<?php

declare(strict_types=1);

namespace ProofMark\ShowAd\Tests;

use PHPUnit\Framework\TestCase;
use ProofMark\ShowAd\AccessPolicy\AccessPolicyEvaluator;
use ProofMark\ShowAd\Request\RequestContext;

final class AccessPolicyEvaluatorTest extends TestCase
{
    private function makeRequest(string $ua = '', string $ip = '127.0.0.1', array $headers = []): RequestContext
    {
        $headers = array_merge(['user-agent' => $ua], $headers);
        return new RequestContext(
            'GET',
            '/protected',
            [],
            $headers,
            [],
            $ip,
            'https://example.com/protected'
        );
    }

    public function testAllowsCidrAllowlist(): void
    {
        $evaluator = new AccessPolicyEvaluator();
        $request = $this->makeRequest('Mozilla/5.0', '10.0.0.5');
        $decision = $evaluator->evaluate($request, [
            'allow_cidrs' => ['10.0.0.0/24'],
        ]);

        self::assertSame('allow', $decision['action']);
        self::assertSame('cidr_allowlist', $decision['reason']);
    }

    public function testAllowsIpv6CidrAllowlist(): void
    {
        $evaluator = new AccessPolicyEvaluator();
        $request = $this->makeRequest('Mozilla/5.0', '2001:db8::1');
        $decision = $evaluator->evaluate($request, [
            'allow_cidrs' => ['2001:db8::/32'],
        ]);

        self::assertSame('allow', $decision['action']);
    }

    public function testCrawlerUaAloneNeverAllows(): void
    {
        $evaluator = new AccessPolicyEvaluator();
        $request = $this->makeRequest(
            'Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)',
            '203.0.113.5'
        );

        $decision = $evaluator->evaluate($request, [
            'crawler' => [
                'enabled' => true,
                'family_cidrs' => [
                    'google' => ['66.249.64.0/19'],
                ],
            ],
        ]);

        self::assertSame('continue', $decision['action']);
    }

    public function testCrawlerVerifiedByCidr(): void
    {
        $evaluator = new AccessPolicyEvaluator();
        $request = $this->makeRequest(
            'Mozilla/5.0 (compatible; Googlebot/2.1)',
            '66.249.64.50'
        );

        $decision = $evaluator->evaluate($request, [
            'crawler' => [
                'enabled' => true,
                'family_cidrs' => [
                    'google' => ['66.249.64.0/19'],
                ],
            ],
        ]);

        self::assertSame('allow', $decision['action']);
        self::assertSame('crawler:google', $decision['reason']);
    }

    public function testCrawlerVerifiedByCloudflareHeader(): void
    {
        $evaluator = new AccessPolicyEvaluator();
        $request = $this->makeRequest(
            'Mozilla/5.0 (compatible; bingbot/2.0)',
            '8.8.8.8',
            ['cf-verified-bot' => 'true']
        );

        $decision = $evaluator->evaluate($request, [
            'crawler' => [
                'enabled' => true,
                'allow_cloudflare_verified_bot' => true,
                'family_cidrs' => [],
            ],
        ]);

        self::assertSame('allow', $decision['action']);
        self::assertSame('crawler:bing', $decision['reason']);
    }

    public function testTrustedIpHeaderTakesPriority(): void
    {
        $evaluator = new AccessPolicyEvaluator();
        $request = $this->makeRequest(
            'Mozilla/5.0',
            '127.0.0.1',
            ['x-forwarded-for' => '198.51.100.42, 10.0.0.1']
        );

        $resolved = $evaluator->resolveClientIp($request, ['X-Forwarded-For']);
        self::assertSame('198.51.100.42', $resolved);
    }

    public function testBeforeProtectCallbackCanRedirect(): void
    {
        $evaluator = new AccessPolicyEvaluator();
        $request = $this->makeRequest('Mozilla/5.0', '127.0.0.1');
        $decision = $evaluator->evaluate($request, [
            'before_protect' => static function () {
                return ['action' => 'redirect', 'redirect_url' => 'https://other.example.com'];
            },
        ]);

        self::assertSame('redirect', $decision['action']);
        self::assertSame('https://other.example.com', $decision['redirect_url']);
    }

    public function testBeforeProtectCallbackCanAllow(): void
    {
        $evaluator = new AccessPolicyEvaluator();
        $request = $this->makeRequest('Mozilla/5.0', '127.0.0.1');
        $decision = $evaluator->evaluate($request, [
            'before_protect' => static function () {
                return ['action' => 'allow', 'reason' => 'premium_user'];
            },
        ]);

        self::assertSame('allow', $decision['action']);
        self::assertSame('premium_user', $decision['reason']);
    }

    public function testIpInCidrsHandlesSingleIp(): void
    {
        $evaluator = new AccessPolicyEvaluator();
        self::assertTrue($evaluator->ipInCidrs('192.0.2.5', ['192.0.2.5']));
        self::assertFalse($evaluator->ipInCidrs('192.0.2.6', ['192.0.2.5']));
    }
}
